src/charm.py

```diff
@@ -64,6 +69,30 @@ def _on_config_changed(self, event) -> None:
         self.model.unit.status = MaintenanceStatus(msg)
         self.configure_workload_container(event)
 
+    def _report_visits_by_ip_action(self, event: ActionEvent) -> None:
+        """Handle the report-visits-by-ip action."""
+        results = self._report_visits_by_ip(event)
+        event.set_results({"ips": str(results)})
+
+    def _report_visits_by_ip(self, _) -> list[(int, str)]:
+        """Report requests to nginx grouped and ordered by IP and report action result."""
+        container = self.unit.get_container(CONTAINER_NAME)
+        log_path = "/var/log/nginx/access.log"
+        with container.pull(log_path) as log_file:
+            list_to_search = log_file.readlines()
+        ip_regex = r"[0-9]+(?:\.[0-9]+){3}"
```

Comment:
Something else came to mind;
* what about IPv6?
* What about IP addresses in the logs unrelated to a request? E.g., is there any other way for an IP address to make its way into the logs other than a request coming in?
* Also consider whether we can source the IP regex from somewhere, e.g., some sort of package